src/data/int/log.lean

```diff
@@ -0,0 +1,245 @@
+Copyright (c) 2022 Eric Wieser. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Eric Wieser
+-/
+import data.nat.log
+import algebra.order.floor
+import algebra.field_power
+
+/-!
+# Integer logarithms in a field with respect to a natural base
+
+This file defines two `ℤ`-valued analogs of the logarithm of `n : R` with base `b : ℕ`:
+
+* `int.log b n`: Lower logarithm, or floor **log**. Greatest `k` such that `↑b^k ≤ n`.
+* `int.clog b n`: Upper logarithm, or **c**eil **log**. Least `k` such that `n ≤ ↑b^k`.
+
+Note that `int.log` gives the position of the left-most non-zero digit:
+```lean
+#eval (int.log 10 (0.09 : ℚ), int.log 10 (0.10 : ℚ), int.log 10 (0.11 : ℚ))
+--    (-2,                    -1,                     -1)
+#eval (int.log 10 (9 : ℚ),    int.log 10 (10 : ℚ),   int.log 10 (11 : ℚ))
+--    (0,                      1,                     1)
+```
+which means it can be used for computing digit expansions
+```lean
+import data.fin.vec_notation
+
+def digits (b : ℕ) (q : ℚ) (n : ℕ) : ℕ :=
+⌊q*b^(↑n - int.log b q)⌋₊ % b
+
+#eval digits 10 (1/7) ∘ (coe : fin 8 → ℕ)
+-- ![1, 4, 2, 8, 5, 7, 1, 4]
+```
+
+## Main results
+
+* For `int.log`:
+  * `int.zpow_log_le_self`, `int.lt_zpow_succ_log_self`: the bounds formed by `int.log`,
+    `(b : R) ^ log b r ≤ r < (b : R) ^ (log b r + 1)`.
+  * `int.zpow_le_iff_le_log`: the galois almost-connection
+* For `int.clog`:
+  * `int.zpow_pred_clog_lt_self`, `int.self_le_zpow_clog`: the bounds formed by `int.log`,
+    `(b : R) ^ (clog b r - 1) < r ≤ (b : R) ^ clog b r`.
+  * `int.le_zpow_iff_clog_le`: the galois almost-connection
+* `int.neg_log_inv`: the link between the two definitions
+
+-/
+variables {R : Type*} [linear_ordered_field R] [floor_ring R]
+
+namespace int
+
+/-- The greatest power of `b` such that `b ^ log b r ≤ r`. -/
+def log (b : ℕ) (r : R) : ℤ :=
+if 1 ≤ r then
+  nat.log b ⌊r⌋₊
+else
+  -nat.clog b ⌈r⁻¹⌉₊
+
+lemma log_of_one_le_right (b : ℕ) {r : R} (hr : 1 ≤ r) : log b r = nat.log b ⌊r⌋₊ :=
+if_pos hr
+
+lemma log_of_right_le_one (b : ℕ) {r : R} (hr : r ≤ 1) : log b r = -nat.clog b ⌈r⁻¹⌉₊ :=
+begin
+  obtain rfl | hr := hr.eq_or_lt,
+  { rw [log, if_pos hr, inv_one, nat.ceil_one, nat.floor_one, nat.log_one_right, nat.clog_one_right,
+        int.coe_nat_zero, neg_zero], },
+  { exact if_neg hr.not_le }
+end
+
+@[simp, norm_cast] lemma log_nat_cast (b : ℕ) (n : ℕ) : log b (n : R) = nat.log b n :=
+begin
+  cases n,
+  { simp [log_of_right_le_one _ _, nat.log_zero_right] },
+  { have : 1 ≤ (n.succ : R) := by simp,
+    simp [log_of_one_le_right _ this, ←nat.cast_succ] }
+end
+
+lemma log_of_left_le_one {b : ℕ} (hb : b ≤ 1) (n : R) : log b n = 0 :=
+begin
+  cases le_total 1 n,
+  { rw [log_of_one_le_right _ h, nat.log_of_left_le_one hb, int.coe_nat_zero] },
+  { rw [log_of_right_le_one _ h, nat.clog_of_left_le_one hb, int.coe_nat_zero, neg_zero] },
+end
+
+lemma log_of_right_le_zero {n : R} (hn : n ≤ 0) (b : ℕ) : log b n = 0 :=
```

Comment:
You're missing the `clog` version of those.

Reply:
Do you think it's worth having clog at all?

Reply:
Yes, because it's adjoint to `zpow`, just from the other side.

Reply:
Am I blind? Don't I already have the clog version of both of these?

Reply:
I would swear they weren't there, but I must have been mistaken.